### PR TITLE
Allow specifying number of iterations via --benchmark_min_time.

### PR DESCRIPTION
**Pull request by oontvoo**
Make the flag accept two new suffixes:
 + <integer>x: number of iterations
 + <floag>s: minimum number of seconds.

This matches the internal benchmark API.

### CODE REVIEW

**Comment by oontvoo** [comment]
@dmah42 Let me know if we should rename the flag in this patch too? Thanks!

**Comment by dmah42** [comment]
aside from the broken build, this is also a breaking change.  before it's merged we should do a release and then release right after it's merged with clear documentation about the behaviour change.

i haven't checked the code yet.. but does it also accept the old behaviour (ie if the "s" is missing")?

**Comment by oontvoo** [comment]
> but does it also accept the old behaviour (ie if the "s" is missing")?

Yes - if suffixes are omitted, then we'd assume it's the number of seconds for backward compat

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:144]: https://github.com/google/benchmark/blob/main/CMakeLists.txt#L141 <-- C++11 is the only requirement (except under MSVC).

no std::variant until C++17.

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:184]: this check is wrong. someone could pass in "0x" expecting it to be a minimum of 0 iterations.

Comment (oontvoo) [reply src/benchmark_runner.cc]: I'm not sure I understand why it's wrong.

If someone passes "<anything>x", then we interpret the number as the number of iterations.
Whether it's zero or not doesn't make a difference. (ie., doesn't change the fact that they are specifying the number of iterations and not the time limit). 

This check is just checking to see whether the value of the flag is for iterations or time. (`.index() == 0` is to check which member of the variant is active.  0 is for the first member `int` and `1` is for second member `double`)

**Comment by oontvoo** [comment]
adding tests - please hold off on the review. Will ping when ready. Thanks!

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:162]: if we have `strtonum` we can get an error, but i'm not sure we do.  at least `strtol` provides some range protection (atoi does no such thing).

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:165]: please avoid the word "sanity check"

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:184]: use strtod and check errno?

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:187]: this is dangerous. "00.0", "0.", ".0" are all valid representations of 0.0.

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:198]: so here we accept 0 is an invalid value for time, but we accept it as a valid time on the command line.  maybe we can simplify your error checks by just not allowing 0 as a valid minimum time.

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:201]: use a switch so we don't miss it if someone adds a new tag.

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:206]: why does this go back to the flag if we already parsed it to determine the BenchTimeType?

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:217]: use benchmark CHECK please

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:183]: if this is something we want users to take action on, we need to log at a higher level than VLOG(2).

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc:214]: the current default minimum time is 0.5. shouldn't we keep that?

File: src/benchmark_runner.cc
Comment (dmah42) [inline src/benchmark_runner.cc]: i don't understand why this is necessary. can you add a comment explaining it please?

Comment (oontvoo) [reply src/benchmark_runner.cc]: The BenchmarkName from the RunResult object was created based on the BenchmarkInstance's BenchmarkName.
Hence its min_time could be zero (if the user never calls `Benchmark::MinTime(...))`). 

We want to update that to the value set in the command-line flag, if user specified the flag.

The logic in this is also a bit convoluted:
 - The default value of BenchmarkInstance's min_time is 0.0 
 - The default value of command-line flag min_time is 0.5
 - We update the name to BenchmarkRunner's min_time only if it's neither of the default values. (*)

(*) This is still not very good reasoning. Maybe we want a way to really know whether user overrides the default value or not and only print the value in that case. (If they override the default value with a default, that's fine too - we'd still print it. Eg., if they set --benchmark_min_time=0.0, we should still print it)

WDYT?

File: src/benchmark_runner.cc
Comment (oontvoo) [reply src/benchmark_runner.cc:214]: Ah, good point

File: src/benchmark_runner.cc
Comment (dmah42) [reply src/benchmark_runner.cc]: i think we should be more consistent. whether the benchmark min time was set by the macro or by the commandline flag, it should be captured somewhere that doesn't require posthoc rewriting of history.

ie, what happens if we store `b.min_time` from the macro _or_ the flag? then all the benchmark naming should just work, right?

Comment (dmah42) [reply src/benchmark_runner.cc]: also, do we want to add a suffix for iterations if the user overrode the iterations on the command line?

Comment (oontvoo) [reply src/benchmark_runner.cc]: Ooh, I see - hmm. I've only added this part so that we can test the flag's setting 
(since there's no way to pull the min_time field out of BenchmarkRunner from a test) ...

any suggestion?

Otherwise, maybe can just rewrite the test to test that the flag was parsed correctly?

Comment (dmah42) [reply src/benchmark_runner.cc]: maybe just add the min time to the benchmark runner so we can access it for testing? :)

Comment (oontvoo) [reply src/benchmark_runner.cc]: Done. 

(ended up added a hook to report the config from BenchmarkRunner to the BenchmarkReporter interface, because from a test, we dont have access to BenchmarkRunner objects)

File: src/benchmark.cc
Comment (dmah42) [inline src/benchmark.cc:82]: should this use `kDefaultMinTime`?

Comment (oontvoo) [reply src/benchmark.cc]: Yes. Done!

File: include/benchmark/benchmark.h
Comment (dmah42) [inline include/benchmark/benchmark.h:285]: i was hoping to not have this declared twice.  is there no way to set the default based on `kDefaultMinTime` directly?

or extract the double default from the string version?

Comment (oontvoo) [reply include/benchmark/benchmark.h:284]: kind of yes :) I've moved the double field to benchmark_runner.cc (where it's used) and parse the string field there.

**Review by dmah42 — COMMENTED** [review]
@LebedevRI should also look at this

**Comment by oontvoo** [comment]
Hi @LebedevRI, any thought on this? Thanks!

File: src/benchmark_runner.cc
Comment (LebedevRI) [inline src/benchmark_runner.cc:184]: This makes it look like other suffixes (`ms`?) are supported

File: src/benchmark_runner.cc
Comment (LebedevRI) [inline src/benchmark_runner.cc:193]: Why do we care to restore the `errno`?

File: src/benchmark_runner.cc
Comment (oontvoo) [reply src/benchmark_runner.cc:184]: Yes, we could extend the suffixes support later - but for the first iteration, let's keep it simple. (And I'd like to get the interface in place first)

File: src/benchmark_runner.cc
Comment (oontvoo) [reply src/benchmark_runner.cc:193]: just thought it's generally good to not pollute errno - can remove this if you'd prefer

**Comment by oontvoo** [comment]
Hi, if there's no further objection, can we please get this rolling? Thanks!

**Review by dmah42 — APPROVED** [review]

**Comment by dmah42** [comment]
i'd like @LebedevRI to approve too

File: src/benchmark_runner.cc
Comment (LebedevRI) [inline src/benchmark_runner.cc:189]: This check is present in release builds, right?

File: test/benchmark_min_time_flag_time_test.cc
Comment (LebedevRI) [inline test/benchmark_min_time_flag_time_test.cc:14]: Is there a negative test for `1ms` / `1z` / `1hs`?

**Comment by LebedevRI** [comment]
Consider this rubber-stamped.

**Comment by LebedevRI** [comment]
(also, what about warm-up option, should it also be affected?)

File: src/benchmark_runner.cc
Comment (chfast) [reply src/benchmark_runner.cc]: Why not std::stod()?

**Comment by oontvoo** [comment]
> (also, what about warm-up option, should it also be affected?)

No, this will not affect the warm-up option.
(We could expand that flag, if it's useful? l)

File: src/benchmark_runner.cc
Comment (oontvoo) [reply src/benchmark_runner.cc]: because it didn't have a very good error-reporting interface (see previous iteration of this patch where stod was used)

File: src/benchmark_runner.cc
Comment (dmah42) [reply src/benchmark_runner.cc:189]: no, it acts like `assert` and is disabled in `NDEBUG` builds

File: test/benchmark_min_time_flag_time_test.cc
Comment (oontvoo) [reply test/benchmark_min_time_flag_time_test.cc:14]: Yes - just added a set of death tests for these

**Comment by dmah42** [comment]
some valid linker errors: `min_time_parse_gtest.cc:(.text+0x6ba): undefined reference to `benchmark::internal::ParseBenchMinTime(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'`

**Comment by oontvoo** [comment]
> > > some valid linker errors: `min_time_parse_gtest.cc:(.text+0x6ba): undefined reference to `benchmark::internal::ParseBenchMinTime(std::__cxx11::basic_string<char, std::char_traits, std::allocator > const&)'`
> > 
> > 
> > Hmm, any idea why the definition is not seen?
> > (The function is defined in benchmark_runner.cc - and moreover,t he same code built in other configurations .... what's difference about these "shared" runs? thinking )
> 
> Ok - now the tests all PASSED in Debug builds but FAILED in Release builds (due to "failed to die"). I must be missing something here but the `#ifndef NDEBUG` should have prevented the tests from running in Release mode - why are they still running? (and hence failing?)

Ah, found it! `TEST_BENCHMARK_LIBRARY_HAS_NO_ASSERTIONS` is the needed guard

**Review by dmah42 — APPROVED** [review]

**Comment by dmah42** [comment]
any more comments here before we rebase and squash and merge?

**Comment by LebedevRI** [comment]
> Consider this rubber-stamped.

**Comment by dmah42** [comment]
wonderful. thank you.

**Pull request merged**
